src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,100 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double SIZE_MULTIPLIER = 1.5;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        grow();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index.");
+        }
+        grow();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        verifyIndex(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        verifyIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        verifyIndex(index);
+        final T removedElement = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[size - 1] = null;
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T removedElement = (T) new Object();
+        for (int i = 0; i < size; i++) {
+            if ((elements[i] != null && elements[i].equals(element))
+                    || elements[i] == null && element == null) {
```

Comment:
looks a bit more compact
```suggestion
            if (elements[i] != null ? elements[i].equals(element) : element == null) {
```